### PR TITLE
Add graphql replication endpoints

### LINKED ISSUE #114: GraphQL schema definition for replication protocol

### PR DESCRIPTION
**Pull request by pietgeursen**
Adds graphql endpoints useful for replication.

## 📋 Checklist

- [x] Add tests that cover your changes
- [x] Add this PR to the _Unreleased_ section in `CHANGELOG.md`
- [x] Link this PR to any issues it closes
- [x] New files contain a SPDX license header

---

Closes #114 and #116 

---

If at all possible I'm going to turn PR review comments requesting changes into issues that can be addressed later in small PRs. This PR is huge (Sorry!) and I'm scared of the time it will take to keep it mergable if it drags out.

So maybe let me know what things _must_ be fixed before this PR is merged and what can be dealt with separately.

### CODE REVIEW

**Comment by sandreae** [comment]
Hi @pietgeursen! I'm working on storage methods tomorrow and some of next week. So far I've been working on the db structure and simple inserts and gets,  thinking more from the node internals rather than replication. I'd like to move onto this though 👍 is there somewhere I can see what you are wanting? Or if not, and you feel like writing me a wish-list, I can try to make it happen 😄

File: aquadoggo/src/bin/dump_gql_schema.rs
Comment (pietgeursen) [inline aquadoggo/src/bin/dump_gql_schema.rs:10]: gotta fix that

**Comment by sandreae** [comment]
I'm still working on the storage provider... got quite far with it and wondering if we want an `AuthorStore`. Currently we don't have a nice way to query for what authors exist, I don't think we ever needed it yet.... but if I build that into the storage provider, I wondered if you wanted anything in particular? I know you are gunna be using author aliases, but I'm not sure if this is something very temporary, or if you'd want it storing?

**Comment by pietgeursen** [comment]
My thought at the moment is that author aliases will be very temporary in
an lru cache.

> Currently we don't have a nice way to query for what authors exist,

I think the client part of replication will want this. For mvp we can just
get every author we know about and ask to replicate them.

File: aquadoggo/src/graphql/ping/mod.rs
Comment (pietgeursen) [inline aquadoggo/src/graphql/ping/mod.rs:1]: might as well remove ping as well

File: aquadoggo/src/graphql/replication/context.rs
Comment (pietgeursen) [inline aquadoggo/src/graphql/replication/context.rs:95]: implement this

File: aquadoggo/src/graphql/replication/context.rs
Comment (pietgeursen) [inline aquadoggo/src/graphql/replication/context.rs:95]: Is there a nicer way to do this rather than having to get the author from an `EntrySigned` and the `log_id` and `seq_num` from an `Entry`?

File: aquadoggo/src/graphql/replication/mod.rs
Comment (pietgeursen) [inline aquadoggo/src/graphql/replication/mod.rs:38]: move this

File: aquadoggo/README.md
Comment (pietgeursen) [inline aquadoggo/README.md:83]: This doesn't have to be about only replication now

File: aquadoggo/src/graphql/mod.rs
Comment (pietgeursen) [inline aquadoggo/src/graphql/mod.rs:6]: fix the ordering

File: aquadoggo/src/graphql/replication/aliased_author.rs
Comment (pietgeursen) [inline aquadoggo/src/graphql/replication/aliased_author.rs:3]: fix ordering

File: aquadoggo/src/graphql/replication/author.rs
Comment (pietgeursen) [inline aquadoggo/src/graphql/replication/author.rs:3]: fix ordering

File: aquadoggo/src/graphql/replication/client/mod.rs
Comment (pietgeursen) [inline aquadoggo/src/graphql/replication/client/mod.rs:5]: fix ordering

File: aquadoggo/src/graphql/replication/context.rs
Comment (pietgeursen) [inline aquadoggo/src/graphql/replication/context.rs:5]: fix ordering

File: aquadoggo/src/graphql/replication/context.rs
Comment (sandreae) [reply aquadoggo/src/graphql/replication/context.rs:98]: I just looked at this again, and wonder why we don't change the signature for `get_certificate_pool()` to accept an entry hash as it's parameter? Then you wouldn't need author, log_id, seq_num.

Comment (pietgeursen) [reply aquadoggo/src/graphql/replication/context.rs:98]: Yeah that's a nice idea!

File: aquadoggo/Cargo.toml
Comment (adzialocha) [inline aquadoggo/Cargo.toml:37]: `mockall` shows up in both development and application dependencies, can it be removed in one of them?

File: aquadoggo/src/graphql/replication/single_entry_and_payload.rs
Comment (sandreae) [inline aquadoggo/src/graphql/replication/single_entry_and_payload.rs:53]: Could these two structs `StorageEntry` and `SingleEntryAndPayload` be combined into one? `StorageEntry` is really just a wrapper around an entry and it's encoded operation:

```rust
#[derive(Debug, Clone, PartialEq)]
pub struct StorageEntry {
    entry_signed: EntrySigned,
    operation_encoded: OperationEncoded,
}
```

**Review by sandreae — APPROVED** [review]
Looks great to me, thanks for all this work, it's really amazing to see actual replication functionality in `aquadoggo`, all very exciting!

I had a few questions (some just for myself) and other small things, nothing major. I'd also be quite interested to know if you came across any unexpected pattern when working with `p2panda-rs`. Like were there structs you were missing, or if there would be a better pattern for interacting with our types (more traits maybe). This can happen in a different thread though.

I'm happy for this to be merged, and if you wanna pull out any of these review points into issues that's all good.

Thanks!! :pray: :pray:

File: aquadoggo/src/bin/dump_gql_schema.rs
Comment (adzialocha) [inline aquadoggo/src/bin/dump_gql_schema.rs:5]: If we add a `default-members = ["aquadoggo_cli"]` in the workspace `Cargo.toml` one can still run `cargo run` and it will start the aquadoggo node. Now with two binary targets one receives an error which makes sense, but it would be nice to easily make it work with a default

File: CHANGELOG.md
Comment (adzialocha) [inline CHANGELOG.md:21]: Missing the PR number and a link to it :cat2:

File: aquadoggo/src/graphql/replication/context.rs
Comment (adzialocha) [inline aquadoggo/src/graphql/replication/context.rs:153]: Mini nitpick: Import order should be `std`, external crates, `crate`, `super` :+1:

... aaaand there is a unnecessary newline on the top of the `tests` mod

File: aquadoggo/src/graphql/replication/context.rs
Comment (adzialocha) [reply aquadoggo/src/graphql/replication/context.rs:189]: Looks really cool indeed!

File: aquadoggo/src/graphql/replication/aliased_author.rs
Comment (adzialocha) [inline aquadoggo/src/graphql/replication/aliased_author.rs:8]: I'm not sure if I understand the concept of `AliasedAuthor`, could you maybe add a docstring here explaining it?

File: aquadoggo/src/graphql/replication/client/mod.rs
Comment (adzialocha) [inline aquadoggo/src/graphql/replication/client/mod.rs:15]: Just for understanding, I guess this is a preparation for later: We're generating a `schema.graphql` with an own `bin` from our schema definition in the code, to then re-import it here for `graphql_client`. How and where is this (`GetEntryByHash`) then used?

**Review by adzialocha — CHANGES_REQUESTED** [review]
Woah, thank you so much @pietgeursen, this looks so nice! Also, I've learned some things about mocking in Rust :smiley_cat: 

@sandreae already had some good comments, especially curious to hear about when you used the `p2panda_rs` API and if you have any advise in how we can improve it. Maybe there is also a way to avoid wrapping the types?

Otherwise I mostly have a bunch of nitpicks regarding style (still haven't done my style guide homework ..), also could be good to run `cargo fmt` once before we merge it :+1:

File: aquadoggo/src/graphql/replication/context.rs
Comment (pietgeursen) [reply aquadoggo/src/graphql/replication/context.rs:24]: Yeah, I don't think we'll put the aliases in the `AuthorStore`. 

> but there could be access to the public keys then.

Still not sure if we'll need that in here. TBD

File: aquadoggo/src/graphql/replication/mod.rs
Comment (pietgeursen) [reply aquadoggo/src/graphql/replication/mod.rs:52]: Yeah exactly. 

If you don't use the `PhantomData` type the compiler will complain because you don't actually use the type in the struct.

I did it this way so that `ReplicationRoot` methods didn't have to be generic over `EntryStore`

File: aquadoggo/src/graphql/replication/mod.rs
Comment (pietgeursen) [reply aquadoggo/src/graphql/replication/mod.rs:131]: :tada:

File: aquadoggo/src/graphql/replication/single_entry_and_payload.rs
Comment (pietgeursen) [reply aquadoggo/src/graphql/replication/single_entry_and_payload.rs:53]: No, I intentionally separate them.

`SingleEntryAndPayload` is the type you get back when you do partial replication. It has a gql method that lets you request the `skiplinks`.

I wanted to keep this apart from an "ordinary" entry that would be part of a collection you'd get back from calling eg `get_entries_newer_than_seq`.

It's some defensive programming. Imagine if you could write a query like this:

```gql
query GetEntriesNewerThanSeq($logId: LogId!, $author: Author!, $sequenceNumber: SequenceNumber!, $first: Int, $after: String) {
    getEntriesNewerThanSeq(logId: $logId, author: $author, sequenceNumber: $sequenceNumber, first: $first, after: $after){
        pageInfo {
            hasNextPage
        }
        edges {
            cursor,
            node {
                entry,
                payload,
                skiplinks
            }
        }
    }
}
```

Asking for the skiplinks here would mean it would have to resolve the skip links for every single entry it returns. It would be a handy way to ddos the server.

File: aquadoggo/src/graphql/replication/client/mod.rs
Comment (pietgeursen) [reply aquadoggo/src/graphql/replication/client/mod.rs:15]: > Probably this is a preparation for later

Yeah, sorry, confusing. It really belongs in the next PR.

> How and where is this (GetEntryByHash) then used?

You'll just have to wait and see ;)

File: aquadoggo/src/graphql/replication/single_entry_and_payload.rs
Comment (pietgeursen) [reply aquadoggo/src/graphql/replication/single_entry_and_payload.rs]: Good catch

File: aquadoggo/src/graphql/replication/aliased_author.rs
Comment (pietgeursen) [reply aquadoggo/src/graphql/replication/aliased_author.rs:16]: Added an explanation. Let me know if it's helpful.

File: aquadoggo/src/graphql/replication/testing/mod.rs
Comment (pietgeursen) [reply aquadoggo/src/graphql/replication/testing/mod.rs:13]: Yeah, it's the hoop we have to jump through for the mocking crate to work with an async_trait that uses generics.

File: aquadoggo/src/graphql/replication/aliased_author.rs
Comment (pietgeursen) [reply aquadoggo/src/graphql/replication/aliased_author.rs:16]: I'm still not sure what's gonna happen with the aliased authors. I think I need to re-think a few things around it. But I don't want to get bogged down in it either. I expect it will change.

File: aquadoggo/src/graphql/replication/single_entry_and_payload.rs
Comment (sandreae) [reply aquadoggo/src/graphql/replication/single_entry_and_payload.rs]: Ahh! Thanks for explaining :+1:

File: aquadoggo/src/graphql/replication/entry.rs
Comment (sandreae) [reply aquadoggo/src/graphql/replication/entry.rs:9]: Yeh, +1 on the easy to understand (read), that's a big plus.

**Comment by cafca** [comment]
Reviewing this atm, please hold off on merging.

**Review by cafca — CHANGES_REQUESTED** [review]
Quite exciting to see replication enter the picture! And how fitting that this pr has that nice round number. In my comments I mentioned when things can also be handled in a follow-up. The main thing for me here is that I would like to understand better the approach with generating the schema before hand or maybe at least understand whether this is fundamentally in conflict with the way i've started to implement dynamic queries #141.

File: aquadoggo/src/graphql/replication/single_entry_and_payload.rs
Comment (cafca) [inline aquadoggo/src/graphql/replication/single_entry_and_payload.rs:31]: I think `certificatePool` might be a better name for this.

File: aquadoggo/src/graphql/replication/client/mod.rs
Comment (pietgeursen) [reply aquadoggo/src/graphql/replication/client/mod.rs:15]: The schema is generated so the client can can do code generation. 

See https://github.com/p2panda/aquadoggo/pull/137/files#diff-cd97e251a79c2f076a0eb780a8e92a956e7c9d162350683f7a98c85ca04d8eeb for how I'm gonna use it.

File: aquadoggo/src/lib.rs
Comment (pietgeursen) [reply aquadoggo/src/lib.rs:21]: I made it pub assuming it's going to be useful if / when I get to using qp2p.

File: aquadoggo/src/graphql/replication/single_entry_and_payload.rs
Comment (pietgeursen) [reply aquadoggo/src/graphql/replication/single_entry_and_payload.rs:31]: I don't really like the naming `certificatePool`. Like, what about is it a pool? A pool for me implies that it is unordered. Just my two cents tho.

I've renamed it to certificate pool though.

File: aquadoggo/src/graphql/replication/testing/mod.rs
Comment (pietgeursen) [reply aquadoggo/src/graphql/replication/testing/mod.rs:1]: There aren't tests in here, it's code that's used by the unit tests in `mod.rs` and `context.rs`

File: aquadoggo/src/graphql/context.rs
Comment (pietgeursen) [reply aquadoggo/src/graphql/context.rs:24]: Made an issue for this

File: aquadoggo/README.md
Comment (pietgeursen) [reply aquadoggo/README.md]: I updated the `README` instructions. That ok now?

File: aquadoggo/src/graphql/replication/client/mod.rs
Comment (adzialocha) [reply aquadoggo/src/graphql/replication/client/mod.rs:15]: Thank you for clarifing!

**Review by cafca — APPROVED** [review]

File: aquadoggo/src/graphql/replication/public_key.rs
Comment (sandreae) [reply aquadoggo/src/graphql/replication/public_key.rs:9]: I don't have a strong feeling on this right now, we still need to do some renaming in `p2panda-rs` so maybe we could review this again later?

**Review by adzialocha — APPROVED** [review]
Clippy doesn't seem to be happy, maybe that needs fixing before meeging, otherwise it's good to go! ✌️✨🌟

**Comment by sandreae** [comment]
A few of those clippy errors are coming from `db/store` as some modules there were made public (so need doc strings). I'm happy to fix those in a follow-up PR.

**Comment by pietgeursen** [comment]
I've tidied up all the clippy warnings. I had to add a few doc strings in @cafca's graphql client stuff, make sure you're ok with them.

**Comment by pietgeursen** [comment]
:relieved:

**Pull request merged**
